File: react-school-frontend/src/services/YearService.ts
// yearService.ts
import api from "@/services/api";

export interface Year {
  id: number;
  year: string; // e.g., "1AM", "2AM", "3AS"
  yearType: string;
  label?: string;
}

// Get all years
export const getAllYears = async (): Promise<Year[]> => {
  const { data } = await api.get("/years");
  return data || [];
};

// Get years in sequence (1AM, 2AM, 3AM, 4AM, 1AS, 2AS, 3AS)
export const getYearsInSequence = async (): Promise<Year[]> => {
  const { data } = await api.get("/years/sequence");
  return data || [];
};

// Get middle school years (1AM - 4AM)
export const getMiddleSchoolYears = async (): Promise<Year[]> => {
  const { data } = await api.get("/years/middle-school");
  return data || [];
};

// Get high school years (1AS - 3AS)
export const getHighSchoolYears = async (): Promise<Year[]> => {
  const { data } = await api.get("/years/high-school");
  return data || [];
};

// Get year by label
export const getYearByLabel = async (label: string): Promise<Year | null> => {
  try {
    const { data } = await api.get(`/years/label/${label}`);
    return data || null;
  } catch {
    return null;
  }
};

// Get next year after a specific year label

export const getNextYearAfter = async (yearLabel: string): Promise<Year | null> => {
  try {
    const { data } = await api.get(`/years/${yearLabel}/next`);
    return data || null;
  } catch {
    return null;
  }
};

// Check if year is the last year (3AS)
export const isLastYear = async (yearLabel: string): Promise<boolean> => {
  try {
    const { data } = await api.get(`/years/${yearLabel}/is-last`);
    return data || false;
  } catch {
    return false;
  }
};

// Check if year is the first year (1AM)
export const isFirstYear = async (yearLabel: string): Promise<boolean> => {
  try {
    const { data } = await api.get(`/years/${yearLabel}/is-first`);
    return data || false;
  } catch {
    return false;
  }
};

// Get streams for a year
export const getStreamsByYear = async (yearId: number) => {
  try {
    const { data } = await api.get(`/years/${yearId}/streams`);
    return data || [];
  } catch {
    return [];
  }
};

File: spring-boot-school-backend/demo/src/main/java/com/educationonline/backend/services/YearService.java
package com.educationonline.backend.services;

import com.educationonline.backend.entities.Years;
import com.educationonline.backend.entities.Years.yearValue;
import com.educationonline.backend.repositories.yearRepo;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class YearService {

    private final yearRepo yearRepository;

    /**
    1AM → 2AM → 3AM → 4AM → 1AS → 2AS → 3AS
     */
    public Optional<Years> getNextYear(String currentYearLabel) {
        try {
            yearValue currentYear = yearValue.fromValue(currentYearLabel);
            yearValue nextYear = getNextYearEnum(currentYear);
            
            if (nextYear != null) {
                return yearRepository.findByYear(nextYear);
            }
            return Optional.empty();
        } catch (IllegalArgumentException e) {
            return Optional.empty();
        }
    }

    /**
     * Get next year enum value based on progression
     * 1AM → 2AM → 3AM → 4AM → 1AS → 2AS → 3AS → null
     */
    private yearValue getNextYearEnum(yearValue current) {
        return switch (current) {
            case Y1_AM -> yearValue.Y2_AM;
            case Y2_AM -> yearValue.Y3_AM;
            case Y3_AM -> yearValue.Y4_AM;
            case Y4_AM -> yearValue.Y1_AS;  // Transition from middle to high school
            case Y1_AS -> yearValue.Y2_AS;
            case Y2_AS -> yearValue.Y3_AS;
            case Y3_AS -> null;  // Last year, no next year
        };
    }

    public Optional<Years> getYearByLabel(String label) {
        try {
            yearValue yearVal = yearValue.fromValue(label);
            return yearRepository.findByYear(yearVal);
        } catch (IllegalArgumentException e) {
            return Optional.empty();
        }
    }

    /**
     * Get all years in sequence order
     */
    public List<Years> getAllYearsInOrder() {
        List<yearValue> sequence = List.of(
                yearValue.Y1_AM,
                yearValue.Y2_AM,
                yearValue.Y3_AM,
                yearValue.Y4_AM,
                yearValue.Y1_AS,
                yearValue.Y2_AS,
                yearValue.Y3_AS
        );

        return sequence.stream()
                .flatMap(yearVal -> yearRepository.findByYear(yearVal).stream())
                .toList();
    }

    /**
     * Check if a year is the last year (3AS)
     */
    public boolean isLastYear(String yearLabel) {
        try {
            yearValue year = yearValue.fromValue(yearLabel);
            return year == yearValue.Y3_AS;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    /**
     * Check if a year is the first year (1AM)
     */
    public boolean isFirstYear(String yearLabel) {
        try {
            yearValue year = yearValue.fromValue(yearLabel);
            return year == yearValue.Y1_AM;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    /**
     * Get all middle school years (1AM - 4AM)
     */
    public List<Years> getMiddleSchoolYears() {
        return yearRepository.findAll().stream()
                .filter(y -> y.getYear().isMiddleSchool())
                .toList();
    }

    /**
     * Get all high school years (1AS - 3AS)
     */
    public List<Years> getHighSchoolYears() {
        return yearRepository.findAll().stream()
                .filter(y -> y.getYear().isHighSchool())
                .toList();
    }
}
